### PR TITLE
S3 application settings

### PR DESCRIPTION
**Pull request by CTMBNara**
This PR is based on the work contributed by @muuki88 in PR #2733. We deeply appreciate @muuki88’s initial efforts, which greatly helped in shaping this final implementation. 🙌💪❤️

### CODE REVIEW

**Review by And1sS — DISMISSED** [review]

**Comment by Compile-Ninja** [comment]
@muuki88 When you have time, could you please approve this change? 🙂

File: src/main/java/org/prebid/server/spring/config/SettingsConfiguration.java [hunk 1]
```diff
@@ -217,6 +227,115 @@ public DatabasePeriodicRefreshService ampDatabasePeriodicRefreshService(
...
+        }
+
+        @Bean
+        S3AsyncClient s3AsyncClient(S3ConfigurationProperties s3ConfigurationProperties) throws URISyntaxException {
+            final AwsBasicCredentials credentials = AwsBasicCredentials.create(
+                    s3ConfigurationProperties.getAccessKeyId(),
+                    s3ConfigurationProperties.getSecretAccessKey());
+            final Region awsRegion = Optional.ofNullable(s3ConfigurationProperties.getRegion())
+                    .map(Region::of)
+                    .orElse(Region.AWS_GLOBAL);
```
Comment (muuki88) [inline src/main/java/org/prebid/server/spring/config/SettingsConfiguration.java:282]: I like that coding style 🤩

Still have to get used to the :: syntax in Java.

File: src/test/groovy/org/prebid/server/functional/tests/storage/StoredResponseS3Spec.groovy [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+        assert response.seatbid[0]?.bid?.size() == storedAuctionResponse.bid.size()
+        assert response.seatbid[0]?.bid[0]?.impid == storedAuctionResponse.bid[0].impid
+        assert response.seatbid[0]?.bid[0]?.price == storedAuctionResponse.bid[0].price
+        assert response.seatbid[0]?.bid[0]?.id == storedAuctionResponse.bid[0].id
+
+        and: "PBS not send request to bidder"
+        assert !bidder.getRequestCount(bidRequest.id)
+    }
+
+    @PendingFeature
```
Comment (muuki88) [inline src/test/groovy/org/prebid/server/functional/tests/storage/StoredResponseS3Spec.groovy:43]: This is a neat annotation

File: src/test/java/org/prebid/server/settings/S3ApplicationSettingsTest.java [hunk 3]
```diff
@@ -0,0 +1,404 @@
...
+                        ResponseBytes.fromByteArray(
+                                GetObjectResponse.builder().build(),
+                                "invalidJson".getBytes())));
+
+        // when
+        final Future<Account> result = target.getAccountById("invalidJsonId", timeout);
+
+        // then
+
+        result.onComplete(context.failing(cause -> {
```
Comment (muuki88) [inline src/test/java/org/prebid/server/settings/S3ApplicationSettingsTest.java:160]: Will the test timeout or give a proper error message if it wouldn't fail?

File: src/main/java/org/prebid/server/settings/service/S3PeriodicRefreshService.java [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+
+    @Override
+    public void initialize(Promise<Void> initializePromise) {
+        fetchStoredDataResult(clock.millis(), MetricName.initialize);
+        if (refreshPeriod > 0) {
+            logger.info("Starting s3 periodic refresh for " + cacheType + " every " + refreshPeriod + " s");
+            vertx.setPeriodic(refreshPeriod, ignored -> fetchStoredDataResult(clock.millis(), MetricName.update));
+        }
+
+        initializePromise.tryComplete();
```
Comment (muuki88) [inline src/main/java/org/prebid/server/settings/service/S3PeriodicRefreshService.java:83]: I'm still favouring the solution where the promise is completed with the first fetch result.

Failing to fetch stored impressions/requests is a critical sign that the new instance shouldn't startup. 

And it should also open a port until the initial data is loaded.

**Comment by muuki88** [comment]
I took a look on my smartphone with the Github App 😄

The only thing is the refresh service initialize promise. I'm convinced that it should resolve the promise with the initial fetch result

File: src/test/java/org/prebid/server/settings/S3ApplicationSettingsTest.java (hunk 3 above)
Comment (CTMBNara) [reply src/test/java/org/prebid/server/settings/S3ApplicationSettingsTest.java:160]: Yes.

**Review by muuki88 — APPROVED** [review]

**Comment by muuki88** [comment]
Looks good 🥳 Thanks for polishing this up and make it ready to go.

**Review by Compile-Ninja — APPROVED** [review]

**Pull request merged**